### PR TITLE
add pipeline trigger provider

### PR DESCRIPTION
**Pull request by pipipipipi43**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
add custom pipeline trigger
See the link below for details

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://terminus-org.app.terminus.io/erda/dop/projects/387/issues/all?id=208384&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sImFzc2lnbmVlSURzIjpbIjEwMDA4MDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=TASK)

#### Specified Reviewers:

/assign @sfwn

### CODE REVIEW

File: modules/pipeline/provider.go [hunk 1]
```diff
@@ -25,13 +25,15 @@ import (
...
 	_ "github.com/erda-project/erda/modules/pipeline/aop/plugins"
+	"github.com/erda-project/erda/modules/pipeline/providers/trigger"
 )
 
 type provider struct {
 	CmsService         pb.CmsServiceServer `autowired:"erda.core.pipeline.cms.CmsService"`
 	ReconcilerElection election.Interface  `autowired:"etcd-election@reconciler"`
 	GcElection         election.Interface  `autowired:"etcd-election@gc"`
 	Router             httpserver.Router   `autowired:"http-router"`
+	TriggerService     *trigger.TriggerService `autowired:"erda.core.pipeline.trigger.TriggerService"`
```
Comment (kakj-go) [inline modules/pipeline/provider.go:36]: trigger.TriggerService is an interface type, which itself is a pointer

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+	pipelineTriggers, err := s.triggerDbClient.ListPipelineTriggers(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get trigger label
+	triggerLabel := make(map[string]string)
+	for key, val := range req.Label {
+		triggerLabel[fmt.Sprintf("triggers.%s.%s", req.EventName, key)] = val
```
Comment (kakj-go) [inline modules/pipeline/providers/trigger/trigger.service.go:64]: i think "triggers.%s.%s"  add to pkg/expression package

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	for _, trigger := range pipelineTriggersMap {
+
+		pipelineDefinition, err := s.definitionDbClient.GetPipelineDefinitionByNameAndSource(trigger.PipelineSource, trigger.PipelineYmlName)
+		if err != nil {
+			return nil, err
+		}
+
+		if pipelineDefinition.PipelineYmlName != "" && pipelineDefinition.PipelineYml != "" && pipelineDefinition.PipelineSource != "" {
+			if pipelineDefinition.Extra.CreateRequest == nil {
+				break
```
Comment (kakj-go) [inline modules/pipeline/providers/trigger/trigger.service.go:85]: continue not break

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+				PipelineYml:            pipelineDefinition.PipelineYml,
+				ClusterName:            pipelineDefinition.Extra.CreateRequest.ClusterName,
+				PipelineYmlName:        pipelineDefinition.PipelineYmlName,
+				PipelineSource:         pipelineDefinition.PipelineSource,
+				Labels:                 label,
+				NormalLabels:           pipelineDefinition.Extra.CreateRequest.NormalLabels,
+				ConfigManageNamespaces: pipelineDefinition.Extra.CreateRequest.ConfigManageNamespaces,
+				AutoRunAtOnce:          true,
+				AutoStartCron:          false,
+				ForceRun:               true,
```
Comment (kakj-go) [inline modules/pipeline/providers/trigger/trigger.service.go:106]: ForceRun default false not true

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+	}
+
+	return &pb.PipelineTriggerResponse{PipelineIDs: pipelineIDs}, nil
+}
+
+func (s *TriggerService) RegisterTriggerEvent(definition definition.PipelineDefinitionProcess, yml pipelineyml.PipelineYml) error {
+
+	newPipelineTriggerMap := GetPipelineTriggerMap(yml)
+
+	oldPipelineTriggers, err := s.triggerDbClient.GetPipelineTriggerByID(definition.ID)
```
Comment (kakj-go) [inline modules/pipeline/providers/trigger/trigger.service.go:125]: The method parameter definition itself is the latest
same like db value

Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:125]: done

File: modules/pipeline/providers/trigger/trigger.service.go (hunk 4 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:106]: done

File: modules/pipeline/providers/trigger/trigger.service.go (hunk 3 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:85]: done

File: modules/pipeline/provider.go (hunk 1 above)
Comment (pipipipipi43) [reply modules/pipeline/provider.go:36]: done

File: modules/pipeline/providers/trigger/trigger.service.go (hunk 2 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:64]: done

File: .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+CREATE TABLE `pipeline_triggers` (
+  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT COMMENT '自增id',
+  `event_name` varchar(500) NOT NULL DEFAULT '' COMMENT '触发事件',
```
Comment (sfwn) [inline .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql:19]: `event` 191

File: .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  `pipeline_source` varchar(191) NOT NULL DEFAULT '' COMMENT '来源',
+  `pipeline_yml_name` varchar(191) NOT NULL DEFAULT '' COMMENT '名称',
+  `pipeline_definition_id` bigint(20) NOT NULL DEFAULT '0' COMMENT '定义id',
+  `filter` mediumtext NOT NULL COMMENT '过滤条件',
```
Comment (sfwn) [inline .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql:23]: TODO.
benchtest for 1000 triggers. Add a screenshot.

File: modules/pipeline/providers/trigger/db/op.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+func (client *Client) ListPipelineTriggers(req *pb.PipelineTriggerRequest, ops ...mysqlxorm.SessionOption) ([]PipelineTrigger, error) {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	var triggers []PipelineTrigger
+	if err := session.In("event_name", req.EventName).Find(&triggers); err != nil {
+		return nil, err
+	}
+	FilterTriggers, err := FilterByEvent(triggers, req.Label)
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/db/op.go:64]: filterTriggers

File: modules/pipeline/providers/trigger/db/op.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+			if filterVal != v {
+				isFilterTrigger = false
+				break
+			}
+		}
+		if isFilterTrigger == true {
+			FilterTriggers = append(FilterTriggers, trigger)
+		}
+	}
+	return FilterTriggers, nil
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/db/op.go:101]: filterxxxx

File: modules/pipeline/providers/trigger/db/table.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"time"
+
+	"github.com/erda-project/erda/apistructs"
+)
+
+// PipelineBase represents `pipeline_triggers` table.
+type PipelineTrigger struct {
+	ID                   uint64                    `json:"id" xorm:"pk autoincr"`
+	EventName            string                    `json:"eventName" xorm:"event_name"`
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/db/table.go:26]: event

File: modules/pipeline/providers/trigger/provider.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+	p.triggerService = &TriggerService{
+		p:                  p,
+		triggerDbClient:    &triggerDb.Client{Interface: p.MySQL},
+		definitionDbClient: &definitionDb.Client{Interface: p.MySQL},
+	}
+	if p.Register != nil {
+		pb.RegisterTriggerServiceImp(p.Register, p.triggerService)
+	}
+	definition.RegisterDefinitionHandler(p.triggerService.RegisterTriggerEvent)
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/provider.go:56]: change RegisterTriggerEvent

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+		if err != nil {
+			return nil, err
+		}
+		pipelineIDs = append(pipelineIDs, pipeline.ID)
+	}
+
+	return &pb.PipelineTriggerResponse{PipelineIDs: pipelineIDs}, nil
+}
+
+func (s *TriggerService) RegisterTriggerEvent(definition definition.PipelineDefinitionProcess, yml pipelineyml.PipelineYml) error {
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/trigger.service.go:119]: TODO 
pass original pipeline yml

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+		}
+	}
+
+	if definition.IsDelete {
+		return nil
+	}
+
+	for eventName, newPipelineTrigger := range newPipelineTriggerMap {
+		err := s.triggerDbClient.CreatePipelineTrigger(&triggerDb.PipelineTrigger{
+			ID:                   component_key.GetKey(0),
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/trigger.service.go:142]: uuid.SnowFlakeID

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+	return nil
+}
+
+func (s *TriggerService) checkPermission(ctx context.Context) error {
+	userID := apis.GetUserID(ctx)
+	if userID == "" {
+		return errors.Errorf("failed to get permission(User-ID is empty)")
+	}
+	ordID := apis.GetOrgID(ctx)
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/trigger.service.go:162]: ord

File: modules/pipeline/providers/trigger/trigger.service.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+	if ordID == "" {
+		return errors.Errorf("failed to get permission(org-ID is empty)")
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/trigger.service.go:164]: Org-ID

Comment (sfwn) [reply modules/pipeline/providers/trigger/trigger.service.go:164]: permission (

File: modules/pipeline/providers/trigger/type.go [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+package trigger
+
+import (
+	"context"
+
+	"github.com/erda-project/erda-proto-go/core/pipeline/trigger/pb"
+)
+
+type ConfigManager interface {
```
Comment (sfwn) [inline modules/pipeline/providers/trigger/type.go:23]: remove it

**Review by sfwn — CHANGES_REQUESTED** [review]

File: .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql (hunk 6 above)
Comment (pipipipipi43) [reply .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql:19]: done

File: modules/pipeline/providers/trigger/db/op.go (hunk 8 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/db/op.go:64]: done

File: modules/pipeline/providers/trigger/db/op.go (hunk 9 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/db/op.go:101]: done

File: modules/pipeline/providers/trigger/db/table.go (hunk 10 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/db/table.go:26]: done

File: modules/pipeline/providers/trigger/trigger.service.go (hunk 14 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:162]: done

File: modules/pipeline/providers/trigger/trigger.service.go (hunk 15 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:164]: done

File: modules/pipeline/providers/trigger/type.go (hunk 16 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/type.go:23]: done

File: modules/pipeline/providers/trigger/trigger.service.go (hunk 12 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:119]: done

File: modules/pipeline/providers/trigger/trigger.service.go (hunk 13 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/trigger.service.go:142]: done

File: modules/pipeline/providers/trigger/provider.go (hunk 11 above)
Comment (pipipipipi43) [reply modules/pipeline/providers/trigger/provider.go:56]: done

File: .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql (hunk 7 above)
Comment (pipipipipi43) [reply .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql:23]: Running time of 1000 triggers
ssss

Comment (pipipipipi43) [reply .erda/migrations/pipeline/20210908-create-pipeline-triggers.sql:23]: todo commit in modules/pipeline/providers/trigger/db/table.go

**Review by sfwn — APPROVED** [review]

**Pull request merged**